### ISSUE TITLE
Random 403 errors with apache

### ISSUE THREAD

**Issue by liayn** [docker/for-win#5793]
- [x] I have tried with the latest version of my channel (Stable or Edge)
  - [ ] I have uploaded Diagnostics
  - Diagnostics ID:

### Expected behavior

Accessing a website running in the container works without issues. All requests are served as expected.

### Actual behavior

Random 403 errors on existing files. Mainly assets (css, js) loaded by the website.

### Information

This worked flawlessly in 2.1.x.x.

My setup is a container based on `php:7.2-apache-stretch`.
It uses two mounts:
- `./html:/var/www/html`
- `./some_local_path:/var/www/typo3_src`

A quick tests with `ab` requesting a single JS file:

```
$ ab -n 99 -c 2 http://localhost/typo3/sysext/core/Resources/Public/JavaScript/Contrib/require.js
...
Complete requests:      99
Failed requests:        88
   (Connect: 0, Receive: 0, Length: 88, Exceptions: 0)
Non-2xx responses:      16
...
```
No matter how often I repeat the test, the "Failed requests" number varies between 7 and 98, the "Non-2xx responses" number is pretty stable around 15.

Running the test again with `-c 1` shows **NO** failures at all.

It seems there is a major issues with parallel file access.

Interesting observation: Moving the `require.js` file to document root (`http://localhost/require.js`) and repeating the tests, again shows **NO** failures.

The important difference seems to be that `/var/www/html/typo3` is a **symlink** to `/var/www/typo3_src/typo3/`.

Note:
It does not matter whether the same file is accessed rapidly ("Slow 3G" is "fast" enough) or if multiple different files are accessed fast enough. Seems like the symlink is the bottleneck.

  - Windows Version: Win 10 Pro (10.0.18363 Build 18363)
  - Docker Desktop Version: 2.2.0.3

### Steps to reproduce the behavior

Generate any scenario that causes massive parallel file access on mounted volumes.

E.g.

  1. Container based on php:7.2-apache-stretch
  2. Mount some local path as doc-root (`/var/www/html`)
  3. Mount some other local path somewhere else (`/var/www/something`)
  4. Create a symlink from `/var/www/html/linked -> /var/www/something/target`
  5. Access 1 (or various) file(s) in parallel

**Comment by liayn** [comment]
Since this makes docker-win entirely unusable for my daily work, I had to downgrade to 2.1.0.5 again.

**Comment by robert-heinig** [comment]
Thanks for creating this issue liayn.

I experienced the same problem and have some additional information:
This problem was introduced with 2.2.0.3 and does not occur with 2.2.0.0. I'm using 2.2.0.0 at the moment and everything is fine with this version.
It's not limited to Apache - file access via PHP also failed randomly with 2.2.0.3.

This may indeed be related to symlinks - an older TYPO3 without symlinks did not have this problem with version 2.2.0.3.
In my case the symlink is not mounted separately, it was created within the container by the container itself and refers to a not mounted target directory within the container.

**Comment by ktallafus** [comment]
Yes thank you for opening this issue too. I have this since 2.2.0.0 and it got much worse with the latest update. Makes it impossible to work with. Website frontends working fine, but admin interfaces (eg. TYPO3) with many requests does not work. It get´s a bit better, if i try to throttle the network speed in dev tools but still many requests fail with 403. The same files load fine, when requested manually and one after another. 

The shared volumes are on an SSD btw.

**Comment by jbr-proton** [comment]
Thanks for opening the issue !
Having the exact same thing here. Can't say if it appeared with 2.2.0.0 or not, but it's definitely a problem on 2.2.0.3. 
Only solution found for now is to downgrade. Hoping for an update fot this ! :)

**Comment by liayn** [comment]
@stephen-turner If you could spare a minute to take a look at our problem here? It would be very kind of you. Thanks in advance.

**Comment by stephen-turner** [comment]
If it's broken for symlinks, this is almost certainly a dupe of #5582.

**Comment by liayn** [comment]
It is *NOT* a NTFS junction, it is a symlink in Linux

**Comment by stephen-turner** [comment]
I realise it's a symlink not a junction, but maybe I misunderstood: is it a Windows symlink on the host, or is it just a Linux symlink within the container's (non-mounted) filesystem?

**Comment by liayn** [comment]
It is a symlink created within the container pointing into a mounted filesystem.
In my case the symlink itself is also stored in a mounted FS.

**Comment by robert-heinig** [comment]
No improvements with 2.2.0.4 in this area.

**Comment by djs55** [comment]
Thanks for your report. To anyone who can reproduce this one fairly easily, if you have the time could you reproduce it again and upload a diagnostics and quote the id -- hopefully the error will be in the logs.

Unfortunately the recent symlink improvements were for Windows native symlinks :( When a symlink is created inside Linux we can't guarantee to be able to create a native symlink on the host since it requires some elevated permissions. Therefore we use [mfsymlinks](https://wiki.samba.org/index.php/UNIX_Extensions#Minshall.2BFrench_symlinks) (like we did with Samba) where the symlink is a regular file on Windows with a special header and the target. When Linux calls `stat` and `readlink` on the symlink file we recognise it and return the symlink link metadata. If this kind of symlink is implicated in the repro it suggests the bug is in this codepath -- I'll take a look.

Sorry for the delay in fixing this but thanks again for your feedback.

**Comment by liayn** [comment]
@djs55 
> If this kind of symlink is implicated in the repro it suggests the bug is in this codepath -- I'll take a look.
I can assure you that we are talking exactly about this kind of symlink.

**Comment by robert-heinig** [comment]
@djs55 
Thanks for your response, and yes, the Diagnostic ID is 97F2F41F-CD34-44A2-BE0E-57BA54531B0F/20200315103021

When I delete the log files and Apache returns 403 for existing files, there are entries in the log like
`[11:37:48.375][GoBackendProcess  ][Error  ] time="2020-03-15T11:37:48+01:00" msg="returning EBUSY due to open C:\\[...]\\web\\typo3: The process cannot access the file because it is being used by another process."`
and
`[11:37:48.407][GoBackendProcess  ][Error  ] time="2020-03-15T11:37:48+01:00" msg="Unable to map error to Linux FUSE errno: The process cannot access the file because it is being used by another process. (type was syscall.Errno)"`
I've removed the full path. The path mentioned in the error log entry is a symlink created by the container, the target is within the container. A directory in the directory hierarchy is mounted with Docker, so both source and target of the symlink are accessible within Windows.

Hope this helps.
I could try to create a reduced example if you need a test setup.

**Comment by liayn** [comment]
@djs55 @stephen-turner Sorry for nagging, but do you have any progress on this severe issue for us?

**Comment by djs55** [comment]
I found and fixed some suspicious code in the `mfsymlink` path. If you'd like to try it I've made an experimental build here: https://download-stage.docker.com/win/stable/43700/Docker%20Desktop%20Installer.exe -- let me know if that helps at all.

**Comment by liayn** [comment]
@djs55 Thanks David for your work. I gave it a shot, but unfortunately no improvements here.
Can you reproduce this locally at your machine? Are the reproduction steps outlined enough for you?
Let me know if you need any information in that regard.

**Comment by djs55** [comment]
@liayn from the repro steps and logs I created a slimmed-down test case which starts 2 threads which both attempt to read the same directory through the symlink. This failed pretty quickly on 2.2.0.4. I found another place in the `mfsymlinks` codepath where there was some suspicious code and fixed it. If you'd like to try I've made another experimental build here: https://download-stage.docker.com/win/stable/43737/Docker%20Desktop%20Installer.exe -- let me know if that helps at all.

**Comment by liayn** [comment]
@djs55  
🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 
You are a master!
Works perfectly now and I can confirm a speedup compared to 2.1.0.5 from 5% to 20% depending on the test scenario. Very Nice.
Looking forward to a stable release.

**Comment by robert-heinig** [comment]
Works also for me, thanks a lot @djs55 !

**Comment by liayn** [comment]
Closing this as this got fixed with the latest release 2.2.0.5. Thank you.
